app/src/org/commcare/update/UpdateWorker.kt

```diff
@@ -0,0 +1,105 @@
+package org.commcare.update
+
+import android.content.Context
+import androidx.work.CoroutineWorker
+import androidx.work.WorkerParameters
+import androidx.work.workDataOf
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import org.commcare.CommCareApplication
+import org.commcare.engine.resource.AppInstallStatus
+import org.commcare.engine.resource.ResourceInstallUtils
+import org.commcare.resources.model.InstallCancelled
+import org.commcare.tasks.ResultAndError
+
+/**
+ * Used to stage an update for the seated app in the background. Does not perform
+ * actual update.
+ *
+ */
+class UpdateWorker(appContext: Context, workerParams: WorkerParameters)
+    : CoroutineWorker(appContext, workerParams), InstallCancelled, UpdateProgressListener {
+
+    companion object {
+        const val Progress_Complete = "complete"
+        const val Progress_Total = "total"
+    }
+
+    private lateinit var updateHelper: UpdateHelper
+
+    override suspend fun doWork(): Result {
+
+        updateHelper = UpdateHelper.getNewInstance(true, this, this)
+
+        return coroutineScope {
+            val job = async {
+                doUpdateWork()
+            }
+
+            job.invokeOnCompletion { exception: Throwable? ->
+                when (exception) {
+                    is CancellationException -> {
+                        handleUpdateResult(ResultAndError(AppInstallStatus.Cancelled))
+                    }
+                    else -> {
+                        handleUpdateResult(ResultAndError(AppInstallStatus.UnknownFailure))
+                    }
+                }
+            }
+
+            job.await()
+        }
+
+    }
+
+    private fun doUpdateWork(): Result {
+        var updateResult: ResultAndError<AppInstallStatus>
+
+        // skip if - An update task is already running | no app is seated | user session is not active
+        if (UpdateTask.getRunningInstance() == null &&
+                CommCareApplication.instance().getCurrentApp() != null &&
+                CommCareApplication.instance().getSession().isActive() &&
+                UpdateHelper.shouldAutoUpdate()) {
+
+            updateHelper.startPinnedNotification(CommCareApplication.instance())
+            updateResult = updateHelper.update(ResourceInstallUtils.getDefaultProfileRef())
+        } else {
+            return Result.success()
+        }
+        return handleUpdateResult(updateResult)
+    }
+
+    private fun handleUpdateResult(updateResult: ResultAndError<AppInstallStatus>): Result {
+
+        if (updateResult.data == AppInstallStatus.Cancelled) {
+            updateHelper.OnUpdateCancelled()
+        }
+
+        updateHelper.OnUpdateComplete(updateResult)
+
+        cleanUp()
+
+        return when (updateResult.data.isUpdateInCompletedState) {
+            true -> Result.success()
+            else -> Result.retry()
```

Comment:
I'm not sure this resonates with me.

I think we only want to be applying the backoff-controlled-retries if there was a network reason for the update to fail. The UnknownError state, for example, doesn't feel like it should necessarily be triggering up to (5 min, 25 min, 10 hr) 3 retries over the course of the day before the user is logged out.

Reply:
Makes sense. I was trying to immitate the earlier behaviour but agree that we should only retry on network errors.

Reply:
Got it, that makes sense. Thanks for the revision.